### PR TITLE
[xpu] Fix OneDNN SDPA with GQA + broadcasted mask

### LINKED ISSUE #4369: sdpa meet partition compile failed runtime error when using enable_gqa+attn_mask!=None
### 🐛 Describe the bug

When running doge model, I meet partition compile error for sdpa: SmallDoge/Doge-20M triggers XPU SDPA failure with q=(1,2,128,128), k/v=(1,1,128,128), float16, enable_gqa=True, and a non-contiguous float16 attn_mask of shape (1,2,128,128) with stride (0,0,0,1). The same metadata is reproduced in a PyTorch-only script without Transformers.

```
import sys
import traceback

import torch
import torch.nn.functional as F


def manual_eager_attention(query: torch.Tensor, key: torch.Tensor, value: torch.Tensor) -> torch.Tensor:
    scale = query.size(-1) ** -0.5
    scores = torch.matmul(query, key.transpose(-2, -1)) * scale
    causal_mask = torch.ones(scores.shape[-2:], device=scores.device, dtype=torch.bool).tril()
    scores = scores.masked_fill(~causal_mask, torch.finfo(scores.dtype).min)
    probs = torch.softmax(scores, dim=-1)
    return torch.matmul(probs, value)


def main() -> int:
    device = torch.device("xpu")
    batch_size = 1
    query_heads = 2
    key_value_heads = 1
    seq_len = 128
    head_dim = 128
    dtype = torch.float16

    torch.manual_seed(0)
    query_base = torch.randn(batch_size, seq_len, query_heads, head_dim, device=device, dtype=dtype)
    query = query_base.transpose(1, 2)
    key = torch.randn(batch_size, key_value_heads, seq_len, head_dim, device=device, dtype=dtype)
    value = torch.randn(batch_size, key_value_heads, seq_len, head_dim, device=device, dtype=dtype)
    attn_mask_base = torch.linspace(0.1766357421875, 0.73828125, seq_len, device=device, dtype=dtype)
    attn_mask = torch.as_strided(attn_mask_base, (batch_size, query_heads, seq_len, seq_len), (0, 0, 0, 1))
    scale = head_dim**-0.5

    print("\nRunning manual eager attention control...")
    manual = manual_eager_attention(query, key.expand_as(query), value.expand_as(query))
    torch.xpu.synchronize()
    print(f"manual eager ok: output_shape={tuple(manual.shape)}")

    print("\nRunning torch.nn.functional.scaled_dot_product_attention repro...")
    try:
        output = F.scaled_dot_product_attention(
            query,
            key,
            value,
            attn_mask=attn_mask,
            dropout_p=0.0,
            is_causal=False,
            scale=scale,
            enable_gqa=True,
        )
        torch.xpu.synchronize()
    except Exception:
        print("SDPA failed:")
        traceback.print_exc()
        return 1

    print(f"SDPA ok: output_shape={tuple(output.shape)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
```
on cuda this example is OK.

### Versions

```
PyTorch version: 2.12.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-22-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20250302
Intel GPU driver version:
* intel-opencl-icd:     26.14.37833.4-1~24.04~ppa1
* libze1:       1.28.6-1~24.04~ppa1
Intel GPU models onboard:
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-0d00-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, memory_clock_rate=2800MHz, memory_bus_width=64-bit, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [1] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-4800-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, memory_clock_rate=2800MHz, memory_bus_width=64-bit, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [2] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-6200-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, memory_clock_rate=2800MHz, memory_bus_width=64-bit, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [3] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-b000-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, memory_clock_rate=2800MHz, memory_bus_width=64-bit, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  128
On-line CPU(s) list:                     0-127
Vendor ID:                               GenuineIntel
BIOS Vendor ID:                          Intel(R) Corporation
Model name:                              Intel(R) Xeon(R) 696X
BIOS Model name:                         Intel(R) Xeon(R) 696X  CPU @ 2.4GHz
BIOS CPU family:                         179
CPU family:                              6
Model:                                   173
Thread(s) per core:                      2
Core(s) per socket:                      64
Socket(s):                               1
Stepping:                                1
CPU(s) scaling MHz:                      24%
CPU max MHz:                             4800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               3 MiB (64 instances)
L1i cache:                               4 MiB (64 instances)
L2 cache:                                128 MiB (64 instances)
L3 cache:                                336 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-127
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS Not affected; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] apollo-torch==1.0.3
[pip3] dpcpp-cpp-rt==2025.3.2
[pip3] galore-torch==1.0
[pip3] impi-rt==2021.17.2
[pip3] intel-cmplr-lib-rt==2025.3.2
[pip3] intel-cmplr-lib-ur==2025.3.2
[pip3] intel-cmplr-lic-rt==2025.3.2
[pip3] intel-opencl-rt==2025.3.2
[pip3] intel-openmp==2025.3.2
[pip3] intel-pti==0.16.0
[pip3] intel-sycl-rt==2025.3.2
[pip3] mkl==2025.3.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] oneccl==2021.17.2
[pip3] oneccl-devel==2021.17.2
[pip3] onemkl-license==2025.3.1
[pip3] onemkl-sycl-blas==2025.3.1
[pip3] onemkl-sycl-dft==2025.3.1
[pip3] onemkl-sycl-lapack==2025.3.1
[pip3] onemkl-sycl-rng==2025.3.1
[pip3] onemkl-sycl-sparse==2025.3.1
[pip3] onnx==1.21.0
[pip3] pytorch-msssim==1.0.0
[pip3] tbb==2022.3.1
[pip3] tcmlib==1.4.1
[pip3] torch==2.12.0+xpu
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0+xpu
[pip3] torchcodec==0.13.0+cpu
[pip3] torchdata==0.11.0
[pip3] torchsde==0.2.6
[pip3] torchvision==0.27.0+xpu
[pip3] triton==3.7.0
[pip3] triton-xpu==3.7.1
[pip3] umf==1.0.3
```

### PR DESCRIPTION
**Pull request by LuFinch**
Fix https://github.com/intel/torch-xpu-ops/issues/4369 which run into OneDNN SDPA GQA + broadcasted mask path.

where the `attn_mask_.unsqueeze(2)` should be replaced with `undo_broadcast(reshaped_attn_mask)`
```
if (attn_mask_.has_value() &&
        at::native::onednn::is_broadcast(reshaped_attn_mask)) {
      at::native::onednn::undo_broadcast(reshaped_attn_mask);
    }

    if (num_head_q != num_head_kv) { // Check whether the attention is a
                                     // Grouped-Query Attention (GQA)
      int group_num = num_head_kv;
      int group_size = num_head_q / num_head_kv;
      // oneDNN requires the shape of the query tensor to be represented as
      // [batch_size, num_head_q / num_head_kv, num_head_kv, seq_len_q,
      // head_dim_qk]. Please refer to
      // https://uxlfoundation.github.io/oneDNN/dev_guide_graph_gqa.html#gqa-pattern
      reshaped_query = query_.view(
          {batch_size, group_num, group_size, seq_len_q, head_dim_qk});
      reshaped_key = key_.unsqueeze(2);
      reshaped_value = value_.unsqueeze(2);
      reshaped_attention = attention_.view(
          {batch_size, group_num, group_size, seq_len_q, head_dim_v});
      if (attn_mask_.has_value() && attn_mask_.value().dim() == 4) {
        reshaped_attn_mask = attn_mask_.unsqueeze(2);
      }
    }
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by LuFinch** [comment]
@guangyey @EikanWang  Could you help review?

File: aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp [hunk 1]
```diff
@@ -109,7 +109,7 @@ struct SDPALogicalParams {
       reshaped_attention = attention_.view(
           {batch_size, group_num, group_size, seq_len_q, head_dim_v});
       if (attn_mask_.has_value() && attn_mask_.value().dim() == 4) {
-        reshaped_attn_mask = attn_mask_.value().unsqueeze(2);
+        reshaped_attn_mask = reshaped_attn_mask.unsqueeze(2);
```
Comment (guangyey) [inline aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp:112]: 1. line 105-109 still use the original version, is it fine?
2. add a test case to guard the code change.

Comment (LuFinch) [reply aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp:112]: 1. yes, they are not related to mask.
2. sure, will add

Comment (guangyey) [reply aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp:112]: > yes, they are not related to mask.

I can't understand. `key` and `value` will be `undo_broadcast` in lin 86-91 if they are broadcast. In this situation, in line 107-108, using the original tensor will lose the information of `undo_broadcast`. Why is this fine?

Comment (LuFinch) [reply aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp:112]: Oh, I get your point now. Will fix.

**Review by guangyey — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Comment by LuFinch** [comment]
@guangyey can we merge this PR? The failure is not related.

**Review by guangyey — APPROVED** [review]
